src/bench/checkqueue.cpp

```diff
@@ -60,7 +59,7 @@ static void CCheckQueueSpeedPrevectorJob(benchmark::Bench& bench)
     bench.minEpochIterations(10).batch(BATCH_SIZE * BATCHES).unit("job").run([&] {
         // Make insecure_rand here so that each iteration is identical.
         CCheckQueueControl<PrevectorJob> control(&queue);
-        for (auto vChecks : vBatches) {
+        for (auto&& vChecks : vBatches) {
```

Comment:
3f3479c5be1cd868d51c166a3808565afc56fb9f: This looks wrong? Previously it would create a copy. Now it steals the memory and then continues to iterate on the "empty" memory?

Reply:
[Updated](https://github.com/bitcoin/bitcoin/pull/26749#issuecomment-1383184072).

Reply:
You didn't change anything?? Pretty sure `&` or `&&` are compiled down to the same (wrong) binary.

Otherwise, it would be good if you explained how your change changed anything or provide a benchmark output, etc.

Reply:
> ... then continues to iterate on the "empty" memory?

Why? I don't think that is the case.

```
$ ./src/bench/bench_bitcoin --filter=CCheckQueueSpeedPrevectorJob
...
|              ns/job |               job/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|              161.43 |        6,194,675.97 |    2.4% |      0.06 | `CCheckQueueSpeedPrevectorJob`
```

It is easy to modify `PrevectorJob::operator()` by adding some observable effects, for example a delay or printing out, to verify correctness.

Reply:
`vBatches` is not `const`, so `auto&` or `auto&&` create a mutable reference. Passing this to `Add`, which takes a mutable reference and calls `std::move` on it to cast it to enable move semantics means it will move the memory. `auto` created a copy.

You posted a single bench run, without a run on master there is not much info to be gathered.

Reply:
> `vBatches` is not `const`, so `auto&` or `auto&&` create a mutable reference. Passing this to `Add`, which takes a mutable reference and calls `std::move` on it to cast it to enable move semantics means it will move the memory. `auto` created a copy.

Correct. But it does not imply that the loop '_... continues to iterate on the "empty" memory_', no?

> You posted a single bench run, without a run on master there is not much info to be gathered.

Master branch (f3bc1a72825fe2b51f4bc20e004cef464f05b965):
```
|              ns/job |               job/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|              236.18 |        4,233,975.84 |    1.3% |      0.09 | `CCheckQueueSpeedPrevectorJob`
```

This PR (53b03946e4dad43f10ba2bd869f754555403c097):
```
|              ns/job |               job/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|              171.49 |        5,831,189.43 |    1.7% |      0.07 | `CCheckQueueSpeedPrevectorJob`
```

Reply:
It implies that the state of the vector is unspecified, but valid. (Sometimes it is valid and empty, but not required to be empty). See https://en.cppreference.com/w/cpp/utility/move and the bench output you posted.

Reply:
Right, the state of the local `vBatches` is unspecified after finishing the for-loop. But instances moved into `control` are perfectly valid.

Reply:
The for loop can be run more than once, so they are unspecified *before* starting the loop

Reply:
> The for loop can be run more than once, so they are unspecified _before_ starting the loop

Oh, right... :man_facepalming:

Reply:
Thanks! Fixed.

This PR (cbd2e1c3f9fec7e14c04f5f6ddb4ee41e32e9c9c):
```
|              ns/job |               job/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|              228.13 |        4,383,416.74 |    1.8% |      0.08 | `CCheckQueueSpeedPrevectorJob`
```